s3/lib/bfs/bucket/s3.rb

```diff
@@ -144,6 +144,7 @@ def init_client(**opts)
         config[:credentials] = opts[:credentials] if opts[:credentials]
         config[:credentials] ||= Aws::Credentials.new(opts[:access_key_id].to_s, opts[:secret_access_key].to_s) if opts[:access_key_id]
         config[:credentials] ||= Aws::SharedCredentials.new(profile_name: opts[:profile_name]) if opts[:profile_name]
+        config[:credentials] ||= Aws::AssumeRoleCredentials.new(role_arn: opts[:assume_role], role_session_name: "#{opts[:assume_role]}_#{Time.now.to_i}") if opts[:assume_role]
```

Comment:
So I am not sure that's right and @andyborn and @vrih may correct me on that one, but `assume_role` is a wrapper on top of some other credentials?

I thought you would need to do something like:

```rb
if opts[:assume_role]
  config[:credentials] = Aws::AssumeRoleCredentials.new(
     client: Aws::STS::Client.new(credentials: config[:credentials])
     role_arn: opts[:assume_role], 
     role_session_name: "#{opts[:assume_role]}_#{SecureRandom.urlsafe_base64(12)}",
  )
  ```

Reply:
also, I am not sure if we can auto-generate `role_session_name` or if this is something that must be configurable?

Reply:
I've omitted client here as the docs state:

> If you omit :client option, a new STS::Client object will be constructed.

The ruby docs say role_session_name is a required string. But the cli docs expand this a bit:

> An identifier for the assumed role session.
> 
> Use the role session name to uniquely identify a session when the same role is assumed by different principals or for different reasons. In cross-account scenarios, the role session name is visible to, and can be logged by the account that owns the role. The role session name is also used in the ARN of the assumed role principal. This means that subsequent cross-account API requests that use the temporary security credentials will expose the role session name to the external account in their AWS CloudTrail logs.
> 
> The regex used to validate this parameter is a string of characters consisting of upper- and lower-case alphanumeric characters with no spaces. You can also include underscores or any of the following characters: =,.@-

Reply:
yes, the way i understood it is you need to initialize the STS client with the underlying creds generated from passed key+secret or ENV